diff --git modules/database/src/ioc/rsrv/camessage.c modules/database/src/ioc/rsrv/camessage.c
index ea36ed3..b1ec955 100644
--- modules/database/src/ioc/rsrv/camessage.c
+++ modules/database/src/ioc/rsrv/camessage.c
@@ -117,13 +117,21 @@ static void *casCalloc(size_t count, size_t size)
  *
  * used to be a macro
  */
-static struct channel_in_use *MPTOPCIU (const caHdrLargeArray *mp)
+static struct channel_in_use *MPTOPCIU (const caHdrLargeArray *mp,
+                                       struct client *client)
 {
     struct channel_in_use   *pciu;
     const unsigned      id = mp->m_cid;
 
     LOCK_CLIENTQ;
     pciu = bucketLookupItemUnsignedId (pCaBucket, &id);
+
+    if (pciu && pciu->client != client) {
+        /* reject attempt to use server assigned ID from another client
+         */
+        pciu = NULL;
+    }
+
     UNLOCK_CLIENTQ;
 
     return pciu;
@@ -159,7 +167,7 @@ va_list                 args
     case CA_PROTO_READ_NOTIFY:
     case CA_PROTO_WRITE:
     case CA_PROTO_WRITE_NOTIFY:
-        pciu = MPTOPCIU(curp);
+        pciu = MPTOPCIU(curp, client);
         if(pciu){
             cid = pciu->cid;
         }
@@ -282,7 +290,7 @@ static void log_header (
 
     ipAddrToDottedIP (&client->addr, hostName, sizeof(hostName));
 
-    pciu = MPTOPCIU(mp);
+    pciu = MPTOPCIU(mp, client);
 
     if (pContext) {
         epicsPrintf ("CAS: request from %s => %s\n",
@@ -294,11 +302,6 @@ static void log_header (
 
     epicsPrintf ( "CAS: Request from %s =>   available=0x%x \tN=%u paddr=%p\n",
         hostName, mp->m_available, mnum, (pciu ? (void *)&pciu->dbch : NULL));
-
-    if (mp->m_cmmd==CA_PROTO_WRITE && mp->m_dataType==DBF_STRING && pPayLoad ) {
-        epicsPrintf ( "CAS: Request from %s =>   Wrote string \"%s\"\n",
-        hostName, (char *)pPayLoad );
-    }
 }
 
 /*
@@ -598,7 +601,7 @@ static void read_reply ( void *pArg, struct dbChannel *dbch,
  */
 static int read_action ( caHdrLargeArray *mp, void *pPayloadIn, struct client *pClient )
 {
-    struct channel_in_use *pciu = MPTOPCIU ( mp );
+    struct channel_in_use *pciu = MPTOPCIU ( mp, pClient );
     int readAccess;
     ca_uint32_t payloadSize;
     void *pPayload;
@@ -618,6 +621,9 @@ static int read_action ( caHdrLargeArray *mp, void *pPayloadIn, struct client *p
         SEND_UNLOCK ( pClient );
         return RSRV_ERROR;
     }
+    if (mp->m_count > dbChannelFinalElements(pciu->dbch)) {
+        mp->m_count = dbChannelFinalElements(pciu->dbch);
+    }
 
     payloadSize = dbr_size_n ( mp->m_dataType, mp->m_count );
     status = cas_copy_in_header ( pClient, mp->m_cmmd, payloadSize,
@@ -696,12 +702,16 @@ static int read_notify_action ( caHdrLargeArray *mp, void *pPayload, struct clie
         return RSRV_ERROR;
     }
 
-    pciu = MPTOPCIU ( mp );
+    pciu = MPTOPCIU ( mp, client );
     if ( !pciu ) {
         logBadId ( client, mp, pPayload );
         return RSRV_ERROR;
     }
 
+    if (mp->m_count > dbChannelFinalElements(pciu->dbch)) {
+        mp->m_count = dbChannelFinalElements(pciu->dbch);
+    }
+
     evext.msg = *mp;
     evext.pciu = pciu;
     evext.pdbev = NULL;
@@ -731,13 +741,27 @@ static int write_action ( caHdrLargeArray *mp,
     int                     status;
     long                    dbStatus;
     void                    *asWritePvt;
+    unsigned                size;
 
-    pciu = MPTOPCIU(mp);
+    pciu = MPTOPCIU(mp, client);
     if(!pciu){
         logBadId(client, mp, pPayload);
         return RSRV_ERROR;
     }
 
+    if (INVALID_DB_REQ(mp->m_dataType)) {
+        log_header ("bad put data type", client, mp, pPayload, 0);
+        return RSRV_ERROR;
+    }
+    if (mp->m_count > dbChannelFinalElements(pciu->dbch)) {
+        mp->m_count = dbChannelFinalElements(pciu->dbch);
+    }
+
+    size = dbr_size_n (mp->m_dataType, mp->m_count);
+    if (size > mp->m_postsize) {
+        return RSRV_ERROR;
+    }
+
     if(!rsrvCheckPut(pciu)){
         status = ECA_NOWTACCESS;
         SEND_LOCK(client);
@@ -962,7 +986,6 @@ struct dbChannel *dbch,
 unsigned    cid
 )
 {
-    static unsigned     bucketID;
     unsigned        *pCID;
     struct channel_in_use   *pchannel;
     int         status;
@@ -995,6 +1018,7 @@ unsigned    cid
     LOCK_CLIENTQ;
 
     do {
+        static unsigned     bucketID;
         /*
          * bypass read only warning
          */
@@ -1638,17 +1662,20 @@ static int write_notify_action ( caHdrLargeArray *mp, void *pPayload,
     int status;
     struct channel_in_use *pciu;
 
-    pciu = MPTOPCIU(mp);
+    pciu = MPTOPCIU(mp, client);
     if(!pciu){
         logBadId ( client, mp, pPayload );
         return RSRV_ERROR;
     }
 
-    if (mp->m_dataType > LAST_BUFFER_TYPE) {
+    if (INVALID_DB_REQ(mp->m_dataType)) {
         log_header ("bad put notify data type", client, mp, pPayload, 0);
         putNotifyErrorReply (client, mp, ECA_BADTYPE);
         return RSRV_ERROR;
     }
+    if (mp->m_count > dbChannelFinalElements(pciu->dbch)) {
+        mp->m_count = dbChannelFinalElements(pciu->dbch);
+    }
 
     if(!rsrvCheckPut(pciu)){
         putNotifyErrorReply (client, mp, ECA_NOWTACCESS);
@@ -1656,6 +1683,9 @@ static int write_notify_action ( caHdrLargeArray *mp, void *pPayload,
     }
 
     size = dbr_size_n (mp->m_dataType, mp->m_count);
+    if (size > mp->m_postsize) {
+        return RSRV_ERROR;
+    }
 
     if ( pciu->pPutNotify ) {
 
@@ -1766,11 +1796,11 @@ static int event_add_action (caHdrLargeArray *mp, void *pPayload, struct client
     struct channel_in_use *pciu;
     struct event_ext *pevext;
 
-    if ( INVALID_DB_REQ(mp->m_dataType) ) {
+    if ( INVALID_DB_REQ(mp->m_dataType) || mp->m_postsize < sizeof(*pmi) ) {
         return RSRV_ERROR;
     }
 
-    pciu = MPTOPCIU ( mp );
+    pciu = MPTOPCIU ( mp, client );
     if ( ! pciu ) {
         logBadId ( client, mp, pPayload );
         return RSRV_ERROR;
@@ -1800,6 +1830,10 @@ static int event_add_action (caHdrLargeArray *mp, void *pPayload, struct client
         return RSRV_ERROR;
     }
 
+    if (mp->m_count > dbChannelFinalElements(pciu->dbch)) {
+        mp->m_count = dbChannelFinalElements(pciu->dbch);
+    }
+
     pevext->msg = *mp;
     pevext->pciu = pciu;
     pevext->size = dbr_size_n(mp->m_dataType, mp->m_count);
@@ -1880,8 +1914,8 @@ static int clear_channel_reply ( caHdrLargeArray *mp,
       * Verify the channel
       *
       */
-     pciu = MPTOPCIU(mp);
-     if(pciu?pciu->client!=client:TRUE){
+     pciu = MPTOPCIU(mp, client);
+     if(!pciu){
          logBadId ( client, mp, pPayload );
          return RSRV_ERROR;
      }
@@ -1989,8 +2023,8 @@ static int event_cancel_reply ( caHdrLargeArray *mp, void *pPayload, struct clie
       * Verify the channel
       *
       */
-     pciu = MPTOPCIU(mp);
-     if (pciu?pciu->client!=client:TRUE) {
+     pciu = MPTOPCIU(mp, client);
+     if (!pciu) {
          logBadId ( client, mp, pPayload );
          return RSRV_ERROR;
      }
@@ -2415,6 +2449,14 @@ int camessage ( struct client *client )
             }
             msg.m_postsize  = ntohl ( pLW[0] );
             msg.m_count     = ntohl ( pLW[1] );
+            if (msg.m_postsize >= ca_uint32_max - (sizeof(*mp) + 2 * sizeof ( *pLW ))) {
+                if(CASDEBUG>0) {
+                    errlogPrintf("Client ext msg size too large 0x%08x\n",
+                                 (unsigned)msg.m_postsize);
+                }
+                status = RSRV_ERROR;
+                break;
+            }
             msgsize = msg.m_postsize + sizeof(*mp) + 2 * sizeof ( *pLW );
             pBody = ( void * ) ( pLW + 2 );
         }
@@ -2435,9 +2477,15 @@ int camessage ( struct client *client )
                 SEND_UNLOCK(client);
                 log_header ( "CAS: Client version too old",
                     client, &msg, 0, nmsg );
-                client->recvBytesToDrain = msgsize - bytes_left;
-                client->recv.stk = client->recv.cnt;
-                status = RSRV_OK;
+                if (msgsize >= bytes_left) {
+                    client->recvBytesToDrain = msgsize - bytes_left;
+                    client->recv.stk = client->recv.cnt;
+                    status = RSRV_OK;
+                    break;
+                } else { // msgsize < bytes_left
+                    client->recv.stk += msgsize;
+                    continue;
+                }
             } else {
                 /* silently ignore UDP from old clients */
                 status = RSRV_ERROR;
