diff --git modules/database/src/ioc/db/dbPutNotifyBlocker.cpp modules/database/src/ioc/db/dbPutNotifyBlocker.cpp
index 62fc44612..635012171 100644
--- modules/database/src/ioc/db/dbPutNotifyBlocker.cpp
+++ modules/database/src/ioc/db/dbPutNotifyBlocker.cpp
@@ -164,7 +164,7 @@ void dbPutNotifyBlocker::initiatePutNotify (
         throw cacChannel::outOfBounds();
     }
 
-    if ( type > SHRT_MAX ) {
+    if ( INVALID_DB_REQ(type) ) {
         throw cacChannel::badType();
     }
 
