diff --git configure/os/CONFIG.Common.linuxCommon configure/os/CONFIG.Common.linuxCommon
index 24b1a0c..68159a9 100644
--- configure/os/CONFIG.Common.linuxCommon
+++ configure/os/CONFIG.Common.linuxCommon
@@ -25,13 +25,15 @@ STATIC_LDLIBS_YES= -Wl,-Bdynamic
 
 # Set runtime path for shared libraries if LINKER_USE_RPATH=YES
 SHRLIBDIR_RPATH_LDFLAGS_YES = $(SHRLIB_DEPLIB_DIRS:%=-Wl,-rpath,%)
-SHRLIBDIR_RPATH_LDFLAGS_ORIGIN = $(shell $(MAKERPATH) -O '\$$ORIGIN' -F $(FINAL_DIR) -R $(LINKER_ORIGIN_ROOT) $(SHRLIB_DEPLIB_DIRS))
+SHRLIBDIR_RPATH_LDFLAGS_ORIGIN = $(shell $(MAKERPATH) -O '\$$ORIGIN' -F $(FINAL_DIR) -R $(LINKER_ORIGIN_ROOT) $(SHRLIB_DEPLIB_DIRS)) \
+ $(if $(filter-out 0,$(.SHELLSTATUS)), $(error $(MAKERPATH) failed))
 SHRLIBDIR_LDFLAGS += \
     $(SHRLIBDIR_RPATH_LDFLAGS_$(LINKER_USE_RPATH))
 
 # Set runtime path for products if LINKER_USE_RPATH=YES
 PRODDIR_RPATH_LDFLAGS_YES = $(PROD_DEPLIB_DIRS:%=-Wl,-rpath,%)
-PRODDIR_RPATH_LDFLAGS_ORIGIN = $(shell $(MAKERPATH) -O '\$$ORIGIN' -F $(FINAL_DIR) -R $(LINKER_ORIGIN_ROOT) $(PROD_DEPLIB_DIRS))
+PRODDIR_RPATH_LDFLAGS_ORIGIN = $(shell $(MAKERPATH) -O '\$$ORIGIN' -F $(FINAL_DIR) -R $(LINKER_ORIGIN_ROOT) $(PROD_DEPLIB_DIRS)) \
+ $(if $(filter-out 0,$(.SHELLSTATUS)), $(error $(MAKERPATH) failed))
 PRODDIR_LDFLAGS += \
     $(PRODDIR_RPATH_LDFLAGS_$(LINKER_USE_RPATH))
 
