diff --git modules/libcom/src/misc/cantProceed.c modules/libcom/src/misc/cantProceed.c
index 7afb4750c..19faca45b 100644
--- modules/libcom/src/misc/cantProceed.c
+++ modules/libcom/src/misc/cantProceed.c
@@ -15,6 +15,7 @@
 #include <stdarg.h>
 
 #include "errlog.h"
+#include "envDefs.h"
 #include "cantProceed.h"
 #include "epicsThread.h"
 #include "epicsStackTrace.h"
@@ -53,20 +54,30 @@ LIBCOM_API void * mallocMustSucceed(size_t size, const char *msg)
 
 LIBCOM_API void cantProceed(const char *msg, ...)
 {
+    int shouldAbort = 0;
     va_list pvar;
     va_start(pvar, msg);
     if (msg)
         errlogVprintf(msg, pvar);
     va_end(pvar);
 
-    errlogPrintf(ANSI_RED("CRITICAL ERROR") " Thread %s (%p) can't proceed, suspending.\n",
-            epicsThreadGetNameSelf(), epicsThreadGetIdSelf());
+    envGetBoolConfigParam(&EPICS_ABORT_ON_ASSERT, &shouldAbort);
+
+    errlogPrintf(ANSI_RED("CRITICAL ERROR") " Thread %s (%p) can't proceed, %s.\n",
+            epicsThreadGetNameSelf(), epicsThreadGetIdSelf(), shouldAbort ? "aborting" : "suspending");
 
     epicsStackTrace();
 
     errlogFlush();
 
     epicsThreadSleep(1.0);
-    while (1)
-        epicsThreadSuspendSelf();
+
+    if (shouldAbort) {
+        errlogPrintf("cantProceed() Calling abort()\n");
+        errlogFlush();
+        abort();
+    } else {
+        while (1)
+            epicsThreadSuspendSelf();
+    }
 }
diff --git modules/libcom/src/misc/cantProceed.h modules/libcom/src/misc/cantProceed.h
index 910ace1fe..82cb4af64 100644
--- modules/libcom/src/misc/cantProceed.h
+++ modules/libcom/src/misc/cantProceed.h
@@ -34,12 +34,14 @@
 extern "C" {
 #endif
 
-/** \brief Suspend this thread, the caller cannot continue or return.
+/** \brief Suspend this thread or call abort(), the caller cannot continue
+ *         or return.
  *
  * The effect of calling this is to print the error message followed by
- * the name of the thread that is being suspended. A stack trace will
+ * the name of the thread that can't proceed. A stack trace will
  * also be shown if supported by the OS, and the thread is suspended
- * inside an infinite loop.
+ * inside an infinite loop, or the process aborted, depending on
+ * EPICS_ABORT_ON_ASSERT.
  * \param errorMessage A printf-style error message describing the error.
  * \param ... Any parameters required for the error message.
  */
diff --git modules/libcom/src/osi/epicsMutex.h modules/libcom/src/osi/epicsMutex.h
index 117767a48..2b0e1ec40 100644
--- modules/libcom/src/osi/epicsMutex.h
+++ modules/libcom/src/osi/epicsMutex.h
@@ -41,6 +41,7 @@
 #ifndef epicsMutexh
 #define epicsMutexh
 
+#include "cantProceed.h"
 #include "epicsAssert.h"
 
 #include "libComAPI.h"
@@ -212,9 +213,12 @@ LIBCOM_API epicsMutexLockStatus epicsStdCall epicsMutexLock(
  * This routine does not return if the identifier is invalid.
  * \param ID The mutex identifier.
  **/
-#define epicsMutexMustLock(ID) {                        \
-    epicsMutexLockStatus status = epicsMutexLock(ID);   \
-    assert(status == epicsMutexLockOK);                 \
+#define epicsMutexMustLock(ID) {                                \
+    epicsMutexLockStatus status = epicsMutexLock(ID);           \
+    if(status != epicsMutexLockOK) {                            \
+        cantProceed("epicsMutexMustLock() failed at %s:%d\n",   \
+            __FILE__, __LINE__);                                \
+    }                                                           \
 }
 
 /**\brief Similar to epicsMutexLock() except that the call returns immediately,
diff --git modules/libcom/src/osi/epicsThread.cpp modules/libcom/src/osi/epicsThread.cpp
index 5de8fd99b..c0b4d72ec 100644
--- modules/libcom/src/osi/epicsThread.cpp
+++ modules/libcom/src/osi/epicsThread.cpp
@@ -26,8 +26,8 @@
 #include "epicsAlgorithm.h"
 #include "epicsTime.h"
 #include "epicsThread.h"
-#include "epicsAssert.h"
 #include "epicsGuard.h"
+#include "cantProceed.h"
 #include "errlog.h"
 
 using namespace std;
@@ -345,7 +345,8 @@ extern "C" {
     {
         epicsThreadId id = epicsThreadCreate (
             name, priority, stackSize, funptr, parm );
-        assert ( id );
+        if ( !id )
+            cantProceed("epicsThreadMustCreate: epicsThreadCreate failed\n");
         return id;
     }
 } // extern "C"
diff --git modules/libcom/src/osi/os/RTEMS-score/osdTime.cpp modules/libcom/src/osi/os/RTEMS-score/osdTime.cpp
index b8527d756..36fdeb252 100644
--- modules/libcom/src/osi/os/RTEMS-score/osdTime.cpp
+++ modules/libcom/src/osi/os/RTEMS-score/osdTime.cpp
@@ -15,19 +15,20 @@
 
 #define EPICS_EXPOSE_LIBCOM_MONOTONIC_PRIVATE
 #include <epicsStdio.h>
-#include <rtems.h>
 #include <errno.h>
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/socket.h>
 #include <netinet/in.h>
-#include <rtems/rtems_bsdnet_internal.h>
 #include "epicsTime.h"
 #include "osdTime.h"
 #include "osiNTPTime.h"
 #include "osiClockTime.h"
 #include "generalTimeSup.h"
 
+#include <rtems.h>
+#include <rtems/rtems_bsdnet_internal.h>
+
 extern "C" {
 
 extern rtems_interval rtemsTicksPerSecond;
