diff --git modules/database/src/std/rec/printfRecord.c modules/database/src/std/rec/printfRecord.c
index 72589b2..250e12e 100644
--- modules/database/src/std/rec/printfRecord.c
+++ modules/database/src/std/rec/printfRecord.c
@@ -250,6 +250,8 @@ static void doPrintf(printfRecord *prec)
                             int padding;
 
                             /* Terminate string and measure its length */
+                            if (n > vspace)
+                                n = vspace;
                             pval[n] = 0;
                             added = strlen(pval);
                             padding = width - added;
