diff --git modules/database/src/std/rec/mbbiRecord.c modules/database/src/std/rec/mbbiRecord.c
index 018c861..141c669 100644
--- modules/database/src/std/rec/mbbiRecord.c
+++ modules/database/src/std/rec/mbbiRecord.c
@@ -336,15 +336,15 @@ static void checkAlarms(mbbiRecord *prec, epicsTimeStamp *timeLast)
             alarm = abs((int)floor(afvl));
         }
     }
+    prec->afvl = afvl;
 
     asev = alarm;
     recGblSetSevr(prec, STATE_ALARM, asev);
 
     /* Check for COS alarm */
-    if (val == prec->lalm ||
-        recGblSetSevr(prec, COS_ALARM, prec->cosv))
+    if (val == prec->lalm)
         return;
-
+    recGblSetSevr(prec, COS_ALARM, prec->cosv);
     prec->lalm = val;
 }
 
