diff --git devOpcuaSup/open62541/CONFIG_OPCUA@ devOpcuaSup/open62541/CONFIG_OPCUA@
index 4cf7390..0861b54 100644
--- devOpcuaSup/open62541/CONFIG_OPCUA@
+++ devOpcuaSup/open62541/CONFIG_OPCUA@
@@ -1,6 +1,7 @@
 # OPC UA Configuration for user application
 #==================================================
 # Open62541 client libraries
+_OPCUA := $(dir $(lastword $(MAKEFILE_LIST)))
 
 # include guard
 ifeq (,$(_CONFIG_OPCUA_INCLUDED))
@@ -13,7 +14,7 @@ CLIENT_SUBDIR = open62541
 ifdef T_A
 
 # Path to the Open62541 C++ installation
-OPEN62541 = @OPEN62541@
+OPEN62541 = $(_OPCUA)/../../../vendor
 
 # How the Open62541 shared libraries are deployed
 #   SYSTEM = shared libs are in a system location
@@ -23,8 +24,8 @@ OPEN62541 = @OPEN62541@
 #   EMBED = link Open62541 code statically into libopcua,
 #           the Open62541 libraries are not required on target system
 OPEN62541_DEPLOY_MODE = @OPEN62541_DEPLOY_MODE@
-OPEN62541_LIB_DIR = @OPEN62541_LIB_DIR@
-OPEN62541_SHRLIB_DIR = @OPEN62541_SHRLIB_DIR@
+OPEN62541_LIB_DIR = $(OPEN62541)/lib
+OPEN62541_SHRLIB_DIR = $(OPEN62541_LIB_DIR)
 # How the Open62541 libraries were built
 OPEN62541_USE_CRYPTO = @OPEN62541_USE_CRYPTO@
 OPEN62541_USE_XMLPARSER = @OPEN62541_USE_XMLPARSER@
